### PR TITLE
[CO-336] Implement redemption endpoints

### PR DESCRIPTION
**Pull request by parsonsmatt**
## Description

This PR implements a legacy variant of the redemption internal endpoints.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CO-336

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs [hunk 1]
```diff
@@ -121,6 +122,11 @@ instance Migrate V0.CCoin (V1 Core.Coin) where
 instance Migrate (V1 Core.Coin) V0.CCoin where
     eitherMigrate (V1 c) = pure (V0.encodeCType c)
 
+instance (n ~ m, ValidMnemonicSentence n)
+    => Migrate (Mnemonic n) (V0.CBackupPhrase m) where
+    eitherMigrate =
+        Right . V0.CBackupPhrase
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs]: HOP HOP HOP! Not so fast hahaha :joy: 
Redemption backup phrase are 9-word mnemonics and are distinct from the wallet backup phrases (12-word mnemonics). Even though both turn into a `CBackupPhrase` for backward compatibility, I wonder if it wouldn't be worth it to make this explicit and have two separate instances for `Mnemonic 9` and `Mnemonic 12`. 

Especially when we talked about possibly longer mnemonics for new features that won't be turned into a `CBackupPhrase` at all.

File: wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs [hunk 2]
```diff
@@ -121,6 +122,11 @@ instance Migrate V0.CCoin (V1 Core.Coin) where
 instance Migrate (V1 Core.Coin) V0.CCoin where
     eitherMigrate (V1 c) = pure (V0.encodeCType c)
 
+instance (n ~ m, ValidMnemonicSentence n)
+    => Migrate (Mnemonic n) (V0.CBackupPhrase m) where
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs]: Out of curiosity, any reason to distinguish `n` from `m` here ?

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs [hunk 3]
```diff
@@ -1767,6 +1770,31 @@ instance BuildableSafeGen NodeInfo where
         nfoLocalTimeInformation
         (Map.toList nfoSubscriptionStatus)
 
+-- | A shielded redemption code.
+newtype ShieldedRedemptionCode = ShieldedRedemptionCode
+    { unShieldedRedemptionCode :: Text
+    } deriving (Eq, Show, Generic)
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: What meaning do we give to `Shielded` here? Is that because of the newtype wrapping the basic `Text` type?
If so, I find it a bit confusing ^^" ... What about simply `RedemptionCode` ?

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs [hunk 4]
```diff
@@ -1767,6 +1770,31 @@ instance BuildableSafeGen NodeInfo where
...
+
+-- | The request body for redeeming some Ada.
+data Redemption = Redemption
+    { redemptionWalletId       :: WalletId
+    -- ^ The 'WalletId' that the redemption will go toward.
+    , redemptionAccountIndex   :: AccountIndex
+    -- ^ The index of the account in the wallet that the redemption applies
+    -- to.
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: These (`redemptionWalletId` & `redemptionAccountIndex`) would be moved to function arguments if we indeed proceed with a

`api/v1/wallets/{walletId}/accounts/{accountId}/certificates`

endpoint :+1:

File: wallet/src/Pos/Util/Mnemonic.hs [hunk 5]
```diff
@@ -26,6 +26,9 @@ module Pos.Util.Mnemonic
        , mnemonicToAesKey
        , entropyToMnemonic
        , entropyToByteString
+
+       -- * Reexports
+       , ValidMnemonicSentence
```
Comment (KtorZ) [inline wallet/src/Pos/Util/Mnemonic.hs]: See comment above, I don't think it's needed here. I believe that we do want to treat `Mnemonic 9` and `Mnemonic 12` as separated types throughout the codebase, because they have different semantics in the end.

File: wallet-new/src/Cardano/Wallet/API/V1/Internal/Redemption.hs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+                     migrate)
+import           Cardano.Wallet.API.V1.Types
+import           Pos.Core (TxAux)
+import           Pos.Crypto.Configuration (ProtocolMagic)
+import qualified Pos.Wallet.Web.Methods.Redeem as V0
+import qualified Pos.Wallet.Web.ClientTypes as V0
+import qualified Pos.Util.Servant as V0
+
+type API
+    = Summary "TODO: Describe the redemption endpoint"
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Internal/Redemption.hs]: :+1:

Comment (KtorZ) [reply wallet-new/src/Cardano/Wallet/API/V1/Internal/Redemption.hs]: Note that there's also a combinator `Description` for longer descriptions.

File: wallet-new/src/Cardano/Wallet/API/V1/Internal/Redemption.hs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    -> Redemption
+    -> MonadV1 (WalletResponse Transaction)
+redeemAda pm submitTx r = do
+    let ShieldedRedemptionCode seed = redemptionRedemptionCode r
+        passphrase = maybe def unV1 (redemptionPassphrase r)
+    accountId <- migrate (redemptionWalletId r, redemptionAccountIndex r)
+    let caccountId = V0.encodeCType accountId
+    fmap single . migrate =<< case redemptionMnemonic r of
+        Just mnemonic -> do
+            phrase <- migrate mnemonic
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Internal/Redemption.hs]: There are quite a few places where I didn't do that and simply applied `CBackupPhrase` to the mnemonic to make them backward compatible, completely forgetting about this `migrate` thingy. I wonder how we could enforce the usage of `migrate` :thinking:

File: wallet-new/src/Cardano/Wallet/API/V1/Internal/Redemption.hs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import qualified Pos.Wallet.Web.Methods.Redeem as V0
+import qualified Pos.Wallet.Web.ClientTypes as V0
+import qualified Pos.Util.Servant as V0
+
+type API
+    = Summary "TODO: Describe the redemption endpoint"
+    :> "redemptions"
+    :> "ada"
+    :> ReqBody '[ValidJSON] Redemption
+    :> Get '[ValidJSON] (WalletResponse Transaction)
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Internal/Redemption.hs]: `WalletResponse Transaction` :smile: ?

**Comment by KtorZ** [comment]
+ YT comments about the endpoint shape and location. This code could be located with other account handlers / endpoints imo.

File: wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs (hunk 2 above)
Comment (akegalj) [reply wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs]: I was wondering this as well (on another hand I rarely use this machinery so don't know how it exactly works)

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs (hunk 3 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: The Daedalus codebase refers to it as the shielded redemption code frequently. Or, a shielded vending key:

```
λ ~/Projects/daedalus/ develop ag shieldedRedemptionCode
source/renderer/app/i18n/locales/hr-HR.json
269:  "wallet.redeem.dialog.shieldedRedemptionCodeError": "!!!Invalid shielded vending key",

source/renderer/app/i18n/locales/zh-CN.json
269:  "wallet.redeem.dialog.shieldedRedemptionCodeError": "!!!Invalid shielded vending key",

source/renderer/app/i18n/locales/de-DE.json
269:  "wallet.redeem.dialog.shieldedRedemptionCodeError": "!!!Invalid shielded vending key",

source/renderer/app/i18n/locales/defaultMessages.json
1684:        "id": "wallet.redeem.dialog.shieldedRedemptionCodeError",

source/renderer/app/i18n/locales/ko-KR.json
269:  "wallet.redeem.dialog.shieldedRedemptionCodeError": "!!!Invalid shielded vending key",

source/renderer/app/i18n/locales/ja-JP.json
269:  "wallet.redeem.dialog.shieldedRedemptionCodeError": "無効な保護還元キー",

source/renderer/app/i18n/locales/en-US.json
269:  "wallet.redeem.dialog.shieldedRedemptionCodeError": "Invalid shielded vending key",

source/renderer/app/components/wallet/ada-redemption/AdaRedemptionForm.js
126:    id: 'wallet.redeem.dialog.shieldedRedemptionCodeError',
```

Given that the original datatype in the original wallet just said `seed :: Text -- TODO: newtype this`, I made a best guess and am happy to choose something more meaningful :)

File: wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs (hunk 2 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs]: It's a trick to get better type inference. The `n ~ m` constraint is only checked after the instance is selected. Suppose we have some polymorphic code like:

```
foo :: Mnemonic n -> Int
bar :: CBackupPhrase 9

mnemonic <- convert bar
pure (foo mnemonic)
```

Because `n` is polymorphic here, we can't select the instance `Migrate (Mnemonic n) (CBackupPhrase n)`, because we don't know that these are the same type. So we say "No instance found for `Migrate (Mnemonic n0) (CBackupPhrase 9)`.

When we use the`n ~ m` constraint and two type variables, this allows GHC to select the instance head `Migrate (Mnemonic n) (CBackupPhrase m)`, and after it has selected this instance, it will check the context: `(n ~ m, ValidMnemonicSentence n)`. The `n ~ m` constraint tells GHC that `n` must be equal to `m`, which means that the more specialized type of the instance is `Migrate (Mnemonic m) (CBackupPhrase 9)` and becomes `Migrate (Mnemonic 9) (CBackupPhrase 9)`.

This is a good blog post on the matter: https://chrisdone.com/posts/haskell-constraint-trick

**Comment by parsonsmatt** [comment]
@akegalj ah, fantastic, thanks! That's the info I needed to do the tests :smile:

File: wallet-new/src/Cardano/Wallet/API/V1/Internal/Redemption.hs (hunk 7 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/API/V1/Internal/Redemption.hs]: I'm not sure it's necessary to try and enforce migrate. Sometimes a one-off explicit change is fine, and this is probably one of those instances. But I just didn't know it'd be that easy :sweat_smile:

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs (hunk 4 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: Good call, I have done this :)

File: wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs (hunk 1 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs]: Great catch! I'll create a type family that errors out if it's not 9 or 12 :)

File: wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs [hunk 9]
```diff
@@ -121,7 +122,11 @@ instance Migrate V0.CCoin (V1 Core.Coin) where
 instance Migrate (V1 Core.Coin) V0.CCoin where
     eitherMigrate (V1 c) = pure (V0.encodeCType c)
 
---
+instance (n ~ m, n ~ 12)
+    => Migrate (Mnemonic n) (V0.CBackupPhrase m) where
+    eitherMigrate =
+        Right . V0.CBackupPhrase
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs]: Now seems overkill compare to `Migrate (Mnemonic 12)` :joy:

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs [hunk 10]
```diff
@@ -1,11 +1,13 @@
 {-# LANGUAGE ConstraintKinds            #-}
 {-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingStrategies         #-}
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: Noice :+1:

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs [hunk 11]
```diff
@@ -256,21 +263,46 @@ instance ByteArray.ByteArrayAccess a => ByteArray.ByteArrayAccess (V1 a) where
...
+    ( mw ~ MnemonicWords n
+    , n ~ EntropySize mw
+    , ValidChecksumSize n csz
+    , ValidEntropySize n
+    , ValidMnemonicSentence mw
+    ) => ToSchema (V1 (Mnemonic mw)) where
     declareNamedSchema _ = do
-        NamedSchema _ schm <- declareNamedSchema (Proxy @(Mnemonic 12))
+        NamedSchema _ schm <- declareNamedSchema (Proxy @(Mnemonic mw))
         return $ NamedSchema (Just "V1BackupPhrase") schm
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: What about getting back to plain types and also use a newtype for the backup phrase after all? Having instances on the newtype instead of treating all mnemonics the same way. 
Can do in a different PR though.

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs [hunk 12]
```diff
@@ -1767,6 +1798,86 @@ instance BuildableSafeGen NodeInfo where
...
+
+-- | A shielded redemption code.
+newtype ShieldedRedemptionCode = ShieldedRedemptionCode
+    { unShieldedRedemptionCode :: Text
+    } deriving (Eq, Show, Generic)
+      deriving newtype (ToJSON, FromJSON)
+
+-- | TODO: This instance is probably wrong. What does a shielded redemption
+-- code look like?
+instance Arbitrary ShieldedRedemptionCode where
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: @akegalj An idea maybe ?

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs (hunk 11 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: easy enough to do now :)

File: wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs (hunk 9 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs]: :joy: I'll can it

**Comment by parsonsmatt** [comment]
@KtorZ alright, I've newtyped the other mnemonic. Feel free to merge overnight if you think it's ready.

**Comment by parsonsmatt** [comment]
@KtorZ I am going to merge this once CI finishes. If you want me to change anything else, I can address that in a future PR :smile:

**Comment by KtorZ** [comment]
:+1: 
Only remark: we could remove all those commented out lines in favor of a small note with a commit hash ^.^
